gitlab/client.py

```diff
@@ -1205,6 +1209,21 @@ def _query(
                 error_message="Failed to parse the server message"
             ) from e
 
+        if self._dedupe:
+            duplicate_ids = (
+                set(o["id"] for o in self._data) & self._retrieved_object_ids
```

Comment:
Another issue, as you can see from all the failing tests: we're not actually guaranteed to have `id` attributes returned, some endpoints will return a different attribute as the unique identifier (we use `_id_attr` in our classes for this reason) and `http_list()` could potentially be used to return arbitrary paginated data.

This makes me think it would almost be better to do this in `ListMixin` based on the presence of `self._objc_cls._id_attr`, but not sure if it's too late at that stage to do it efficiently:

https://github.com/python-gitlab/python-gitlab/blob/541a7e3ec3f685eb7c841eeee3be0f1df3d09035/gitlab/mixins.py#L248-L250